subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultInputProperty.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.base.Objects;
+import org.apache.commons.lang.SerializationUtils;
+import org.gradle.internal.io.ClassLoaderObjectInputStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.NotSerializableException;
+import java.io.Serializable;
+
+class DefaultInputProperty implements InputProperty {
+    private transient Object inputProperty;
+    private final byte[] serializedInputProperty;
+
+    private DefaultInputProperty(byte[] serializedInputProperty, Object inputProperty) {
+        this.inputProperty = inputProperty;
+        this.serializedInputProperty = serializedInputProperty;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (obj == this) {
+            return true;
+        }
+        if (obj.getClass() != getClass()) {
+            return false;
+        }
+
+        DefaultInputProperty rhs = (DefaultInputProperty) obj;
+
+        try {
+            tryDeserialize(this, rhs);
+        } catch (IOException e) {
+            // In presence of corruption, both object cannot assert to be equal
+            return false;
```

Comment:
That is a very good idea. I will shelve this for today and will look more in depth how to achieve this.